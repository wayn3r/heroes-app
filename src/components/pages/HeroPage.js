import { useMemo } from 'react'
import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { asset } from '../../helpers/get-asset'
import { getHeroById } from '../../selectors/getHeroById'

const HeroPage = () => {
    const { id } = useParams()
    const hero = useMemo(() => getHeroById(id), [id])
    const navigate = useNavigate()

    if (!hero) return <Navigate to="/" replace />
    const { superhero, publisher, alter_ego, first_appearance, characters } =
        hero
    const handleReturn = () => navigate(-1)
    return (
        <div className="hero-page row mt-5">
            <div className="col-4">
                <img
                    className="hero-page__img animate__animated animate__fadeInLeft"
                    src={asset(`heroes/${id}.jpg`)}
                    alt={superhero}
                />
            </div>

            <div className="col-8 animate__animated animate__fadeIn">
                <h3>{superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item">
                        <b>Alter ego: </b> {alter_ego}
                    </li>
                    <li className="list-group-item">
                        <b>Publisher: </b> {publisher}
                    </li>
                    <li className="list-group-item">
                        <b>Firts appearance: </b> {first_appearance}
                    </li>
                </ul>

                <h5>Characters</h5>
                <p>{characters}</p>

                <button
                    className="btn btn-outline-primary"
                    onClick={handleReturn}
                >
                    &larr; Return
                </button>
            </div>
        </div>
    )
}

export default HeroPage
