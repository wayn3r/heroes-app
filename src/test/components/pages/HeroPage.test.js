import { mount } from 'enzyme'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { act } from '@testing-library/react'
import HeroPage from '../../../components/pages/HeroPage'
const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('Pruebas al componente <HeroPage />', () => {
    const wrapper = mount(
        <MemoryRouter initialEntries={['/hero/marvel-spider']}>
            <Routes>
                <Route path="/hero/:id" element={<HeroPage />} />
                <Route path="/" element={<></>} />
            </Routes>
        </MemoryRouter>
    )
    test('debe mostrar el heroe si hay un heroe en la url', () => {
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.html()).toBeTruthy()
    })

    test('debe volver atras al hacer click en RETURN', () => {
        act(() => {
            wrapper.find('button').prop('onClick')()
        })
        expect(mockNavigate).toHaveBeenCalledWith(-1)
    })

    test('no debe mostrar nada si no hay heroe en la url o no existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/heroe-inexistente123']}>
                <Routes>
                    <Route path="/hero/:id" element={<HeroPage />} />
                    <Route path="/" element={<></>} />
                </Routes>
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.html()).toBeFalsy()
    })
})
