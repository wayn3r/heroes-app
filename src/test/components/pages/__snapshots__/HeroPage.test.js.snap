// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente <HeroPage /> debe mostrar el heroe si hay un heroe en la url 1`] = `
<div
  className="hero-page row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Spider Man"
      className="hero-page__img animate__animated animate__fadeInLeft"
    />
  </div>
  <div
    className="col-8 animate__animated animate__fadeIn"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego: 
        </b>
         
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publisher: 
        </b>
         
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Firts appearance: 
        </b>
         
        Amazing Fantasy #15
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-primary"
      onClick={[Function]}
    >
      ‚Üê Return
    </button>
  </div>
</div>
`;

exports[`Pruebas al componente <HeroPage /> no debe mostrar nada si no hay heroe en la url o no existe 1`] = `""`;
